### PR TITLE
Automatic serialize and deserialize object in collections

### PR DESCRIPTION
**Pull request by ailisp**
Among a few other alternatives (see below), this seems the most developer friendly version of API. See new added test in this PR about how it works. The cons: add about 10KB to contract due to the superserial dependency; need to add a line to deserialize method of the contract to make sure the serializer is properly setup.

Addressed #91 . This approach would also work for other collections, but before adding them let's review and discuss!

The alternatives:
- a manually reviver pattern (or write this pattern / code generation to make it happens automatically): https://stackoverflow.com/questions/14027168/how-to-restore-original-object-type-from-json. Some similar variants:
  - https://clarle.github.io/yui3/yui/docs/json/json-freeze-thaw.html
  - https://rebolon.medium.com/json-reviver-pattern-how-to-restore-object-from-json-into-real-entity-737b1a220450
  - https://riptutorial.com/javascript/example/1389/serializing-and-restoring-class-instances
- use a enhanced JSON serializer library, that has ability to recover class. This PR uses superserial. Alternatives: https://github.com/shaochuancs/esserializer

### CODE REVIEW

**Comment by BenKurrek** [comment]
So with this, all JS contracts would be ~15 $NEAR more expensive than Rust. This isn't a ton and if you're on testnet, it doesn't matter. If you're on mainnet, you probably have the $NEAR since you're deploying a prod level contract.

File: src/collections/lookup-map.ts [hunk 1]
```diff
@@ -1,23 +1,42 @@
...
     }
 
+    getObject(key: any): any {
+        let storageKey = this.keyPrefix + this.serializer.serialize(key)
+        let raw = near.storageRead(storageKey)
+        if (raw !== null) {
+            return this.serializer.deserialize(raw)
+        }
+        return null
+    }
```
Comment (austinabell) [inline src/collections/lookup-map.ts:38]: I don't personally see the reason for having both supported like this. The whole use case for JS SDK is that it is easier to use. I have a strong but low confidence opinion that the lookup map should be typed as `LookupMap<K, V>` where the keys and values would be serialized and have that as the base methods (`get`, `remove`, etc.). If the types are generic, you shouldn't need this superserial library, do you?

If the duplication is kept, I don't see why this wouldn't be swapped where the more primitive ones would be called something else, and the simple methods to use are the name like `get`

File: src/collections/lookup-map.ts [hunk 2]
```diff
@@ -1,23 +1,42 @@
...
 
     containsKey(key: Bytes): boolean {
         let storageKey = this.keyPrefix + key
         return near.storageHasKey(storageKey)
     }
 
+    containsObjectKey(key: any): boolean {
+        let storageKey = this.keyPrefix + this.serializer.serialize(key)
+        return near.storageHasKey(storageKey)
+    }
```
Comment (austinabell) [inline src/collections/lookup-map.ts:24]: If you keep these methods as is, you should probably point the expanded function at the other to make it more maintainable and avoid duplication. Something like
```suggestion
    containsObjectKey(key: any): boolean {
        return this.containsKey(this.serializer.serialize(key))
    }
```

**Comment by volovyk-s** [comment]
I can see why we have such a design here, but anyway:
- I think we should avoid using external libraries in the SDK unless it is absolutely necessary.
- Having two functions for `set`/`get` is not the best DevEx.

I agree with Austin that we should try to implement collections using generics. No guarantees that this is possible, but it is something developers are used to and at the first glance it looks like something that can work.

**Comment by ailisp** [comment]
> So with this, all JS contracts would be ~15 $NEAR more expensive than Rust. This isn't a ton and if you're on testnet, it doesn't matter. If you're on mainnet, you probably have the $NEAR since you're deploying a prod level contract.

No, **without this** standalone JS contract was 5 $NEAR more expensive than Rust (500 KB), This PR adds 10KB, which is 0.1 $NEAR more

**Comment by ailisp** [comment]
@austinabell @volovyk-s Your point makes sense, I struggle in find a good naming.

In rust-sdk, it's `get` (borsh deserialize) and `get_raw` (bytes). In js-sdk, originally we had `get` (bytes). This PR was intended to have this style API: keep `get`, make `get(JSON)Object`, `getBorshObject`
Alternatively:
- make a breaking change to rename `get` to `getRaw`, and make room for new `get` to json deserialize
- also `get` -> `getRaw`, and a get takes an optional argument `options?`. By default, `get` would use JSON deserialize and can be set to use other deserialization method.

**Comment by ailisp** [comment]
> I think we should avoid using external libraries in the SDK unless it is absolutely necessary.

About this, I can try to cherry pick the logic as part of sdk

**Comment by volovyk-s** [comment]
> * make a breaking change to rename `get` to `getRaw`, and make room for new `get` to json deserialize

I think we should not be afraid to make a breaking change here. Consistency between JS and RS versions is more important.

**Comment by ailisp** [comment]
> > I think we should avoid using external libraries in the SDK unless it is absolutely necessary.
> 
> About this, I can try to cherry pick the logic as part of sdk

Updates: it seems not possible to reduce logic from superserial - it completely rewrite instead of reviving on JSON.stringify/parse. Need to copy almost all code from library is not better than just import the library :sweat_smile: . 

After look into superserial's code, it produce JSON for plain object, array and strings but not in all cases. A `TestUser` class of `name` and `age` property is serialized as: `TestUser{"name":"wan2land","age":20}`. It's like a superset of json that still human readable.

We don't have to json serialize the state. For most use case where developer just load/save to state through nearbindgen, it doesn't matter what the format is. So it would be nice to use a lightweight, succinct format, schema included and JS class preserving library to allow fully automatic serialization/deserialization. JSON isn't ideal because it doesn't preserve JS's class. Borsh isn't ideal since it requires schema definition (also recall that the borsh-js support is hacky @volovyk-s ). superserial is so far a reasonable solution, but if you know any more popular library satisfy these requirements would be great:
- lightweight: <= 10KB quickjs bytecode
- succinct: at most as lengthy as JSON, the shorter the better
- schema included: schema is serialized so user can deserialize without specify a spec. Like json. Unlike borsh
- preserve JS class: instances can be deserialized as the original class's instance, instead of object.

**Comment by austinabell** [comment]
> @austinabell @volovyk-s Your point makes sense, I struggle in find a good naming.
> 
> 
> 
> In rust-sdk, it's `get` (borsh deserialize) and `get_raw` (bytes). In js-sdk, originally we had `get` (bytes). This PR was intended to have this style API: keep `get`, make `get(JSON)Object`, `getBorshObject`
> 
> Alternatively:
> 
> - make a breaking change to rename `get` to `getRaw`, and make room for new `get` to json deserialize
> 
> - also `get` -> `getRaw`, and a get takes an optional argument `options?`. By default, `get` would use JSON deserialize and can be set to use other deserialization method.
> 
> - Make `get` a full-capable API: when pass it a key in bytes, it prefix the key and storageRead. When pass it an object key, it auto serialize it, prefix and storageRead

Noting that the duplicate method names just exist with the older version of the SDK collections. There are only the valid types in the one we are migrating to.

**Comment by ailisp** [comment]
Two known issues:
- (bug): for unicode string, superserial may keep it as is, but our collections cannot store `>255` char. A fix would be change how string is serialize and deserialize in superserial: https://github.com/denostack/superserial/blob/4669fa731a74c1f9b6b3b50be5b23cdfedc31d26/serialize.ts#L18

- (enhancement): maybe solve in #89, but if not, the serializer is in every collection and quite annoying to deserialize the serializer itself.

**Review by volovyk-s — APPROVED** [review]
Cool! Sending it to review for DevRel.

**Pull request merged**